### ISSUE TITLE
Base httpexception classes and exception_response function

### ISSUE THREAD

**Issue by evkar** [Pylons/pyramid#1832]
I cannot catch HTTPBadRequest exception in case:

```
try:
    # some code
    raise exception_response(400)
except HTTPBadRequest as e:
    # this code will not be invoked because 
    # exception_response function throws HTTPClientError for 400-status
```
Of cource, i could replace code with:

```
try:
    # some code
    raise HTTPBadRequest()
except HTTPBadRequest as e:
    # catch exception
```

But exception_response function is used by third party code, and i cannot change it. 
Also this bug applies for 500-status errors

**Comment by tseaver** [comment]
Thanks for the report!

`HTTPClientError` is an "abstract base" for all the 4xx-level errors.  Similarly, `HTTPServerError` is a base for all the 5xx errors.  Neither should have a code, and therefore should not be raised when looking up a code in the `status_map`.

**Comment by evkar** [comment]
Thanks for the reply

httpexceptions.py contains next lines (pyramid 1.5.6):

```python
class HTTPClientError(HTTPError):
    """
    base class for the 400s, where the client is in error

    This is an error condition in which the client is presumed to be
    in-error.  This is an expected problem, and thus is not considered
    a bug.  A server-side traceback is not warranted.  Unless specialized,
    this is a '400 Bad Request'
    """
    code = 400
    title = 'Bad Request'
    explanation = ('The server could not comply with the request since '
                   'it is either malformed or otherwise incorrect.')

class HTTPBadRequest(HTTPClientError):
    """
    subclass of :class:`~HTTPClientError`

    This indicates that the body or headers failed validity checks,
    preventing the server from being able to continue processing.

    code: 400, title: Bad Request
    """
    pass
```

Base class has a code equals 400. Unfortunately `HTTPClientError` is raised instead of `HTTPBadRequest` when we are looking up a code in the `status_map`

**Comment by tseaver** [comment]
Fixed via:  #1833, #1834, #1835, #1836

**Comment by evkar** [comment]
Thanks!
